### ISSUE TITLE
The inference results are inconsistent with Huggingface.

### ISSUE THREAD

**Issue by cyfwry** [hao-ai-lab/LookaheadDecoding#30]
Hello! Thanks for this new parallel decoding algorithm.

**Comment by Viol2000** [comment]
Yes, sometimes the results could be more consistent. We owe it to floating point errors, and it is normal -- hf w/ flash-attn and hf w/o flash-attn can also have different outputs sometimes. If you use float32, lade's results should be exactly the same as hf's.

**Comment by cyfwry** [comment]
> Yes, sometimes the results could be inconsistent. We owe it to floating point errors, and it is normal -- hf w/ flash-attn and hf w/o flash-attn can also have different outputs sometimes. If you use float32, lade's results should be exactly the same as hf's.

Thanks for your reply. When I tested, both LookaheadDecoding and HuggingFace used FP16 precision. Moreover, the logic of LookaheadDecoding predicting the next token should be completely consistent with HuggingFace(the token corresponding to ret.out_logits). Why are the results inconsistent?

**Comment by cyfwry** [comment]
> Yes, sometimes the results could be inconsistent. We owe it to floating point errors, and it is normal -- hf w/ flash-attn and hf w/o flash-attn can also have different outputs sometimes. If you use float32, lade's results should be exactly the same as hf's.

Thanks for your reply. I figured this out. In the case of parallel decoding, due to the loss of precision, there is indeed no guarantee that the result will be exactly the same as that of single-step decoding.

**Comment by Viol2000** [comment]
This is the floating point error. Although the logical flows are the same, the computations that happen in GPU are different (i.e., lade computes several tokens per step while hf only computes one token per step).
Different floating point tensor computations could have very similar outputs, but there will always be differences. 
Even if the difference is slight, it will accumulate and turn into inconsistent output someday. It can explain that the inconsistency happens when the output is relatively long.

**Comment by Viol2000** [comment]
And I do not think hf fp16's output is the 'correct' one. hf fp32/lade fp32 outputs should be the 'correct' one. Sometimes, lade fp16's output can align with the fp32 output, while hf fp16's output can be inconsistent with its fp32 output.
